File: src/MessageRoute.java
import java.util.*;

public class MessageRoute {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        int n = in.nextInt();
        int m = in.nextInt();

        int[][] adjList = new int[n+1][n+1];

        for (int i = 0; i < m; i++) {
            int a = in.nextInt();
            int b = in.nextInt();
            adjList[a][b]++;
            adjList[b][a]++;
        }

        Queue<Integer> q = new LinkedList<>();
        q.add(1);
        int[] visited = new int[n+1];
        visited[1] = 1;
        int[] parent = new int[n+1];

        while (!q.isEmpty()) {
            int current = q.poll();

            for (int i = 1; i < adjList.length; i++) {
                if(visited[i] == 0 && adjList[current][i] > 0){
                    visited[i] = 1;
                    parent[i] = current;
                    q.add(i);
                }
            }
        }

        Stack<Integer> route = new Stack<>();
        int current = n;

        if(parent[n] == 0) {
            System.out.println("IMPOSSIBLE");
            return;
        }

        while (current != 0) {
            route.push(current);
            current = parent[current];
        }

        System.out.println(route.size());
        while (!route.isEmpty()) {
            System.out.print(route.pop() + " ");
        }

    }
}

File: src/CSES/MessageRoute.cpp
#include <bits/stdc++.h>
using namespace std;
#define int long long int

int n, m;
vector<vector<int>> adjList;
vector<int> parent;
vector<int> visited;

void bfs()
{
	visited[1] = 1;
	queue<int> q;
	q.push(1);

	while(!q.empty())
	{
		int current = q.front();
		q.pop();
		for(auto neighbor: adjList[current])
		{
			if(visited[neighbor] == 0)
			{
				visited[neighbor] = true;
				parent[neighbor] = current;
				q.push(neighbor);
			}
		}
	} 
}

int32_t main() {

    ios_base::sync_with_stdio(0); 
    cin.tie(0);

    cin >> n >> m;

    adjList.resize(n+1);
    parent.resize(n+1);
    visited.resize(n+1);

    for (int i = 0; i <= n; ++i)
    {
        parent[i] = -1;
    }
    

    for(int i; i < m; ++i) {
        int a, b;
        cin >> a >> b;
        adjList[a].push_back(b);
        adjList[b].push_back(a); 
    }

    bfs();  

    if(parent[n] == -1) {
        cout << "IMPOSSIBLE";
        return 0;
    }

    int current = n;

    stack<int> s;

    while(current != -1) {
        s.push(current);
        current = parent[current];
    }

    cout << s.size() << " \n";

    while(!s.empty()) {
        cout << s.top() << " ";
        s.pop();
    }

}